kuma/users/signal_handlers.py

```diff
@@ -27,6 +37,54 @@ def on_user_signed_up(sender, request, user, **kwargs):
             )
 
 
+@receiver(user_logged_in, dispatch_uid='users.user_logged_in')
+def on_user_logged_in(sender, request, user, **kwargs):
+    sociallogin = kwargs.get('sociallogin')
+    if sociallogin:
+        # Thing is, if someone signs in for the very first time, it'll
+        # trigger two signals: 'user_signed_up' *and* 'user_logged_in'.
+        # If that happens, we only want to send *1* tracking event.
+        # If we listen to both signals we'd get potentially send one
+        # tracking event too many. So, use the `SocialAccount.last_login`
+        # and `SocialAccount.date_joined` to figure out if this was the
+        # the first time.
+        #
+        # Due to how the Django ORM assigns dates, it could be that the
+        # two dates are only different in the number of milliseconds
+        # since the numbers get assigned by calling something like:
+        #
+        #   User.objects.create(
+        #       date_joined=timezone.now(),
+        #       last_login=timezone.now(),
+        #       ...
+        #
+        # NOTE! The SocialAccount (`sociallogin.account` in this context) has
+        # its own `.last_login` and `.date_joined` which is *different* from
+        # the same fields as they're stored in the `auth_user` database.
+        # (peterbe's note): Not sure how that works or why it's so.
+        # So, use the date stored on the user and not on the SocialAccount
+        # instance.
+        if is_almost_same_dates(
```

Comment:
I don't feel entirely comfortable with this approach, for two reasons:
- there is still a chance that the `on_user_logged_in` signal is never fired after the `on_user_signed_up` signal (due to the lack of a verified email on the user account, which I was able to demonstrate locally), so we can't always count on the `on_user_logged_in` to cover both cases
- the `is_almost_same_dates` function doesn't feel bullet-proof

So I tested a different approach that seems simpler and worked for me locally. I was thinking that we should send the `ACTION_SIGN_UP` GA event within the `on_user_signed_up` handler, set a custom attribute on the request object indicating that this was a sign-up (`is_signup` or something like that), and then only send the `ACTION_SIGN_IN` GA event within the `on_user_logged_in` handler if the `request` object does not have the `is_signup` attribute. Something like this:

```python
@receiver(user_signed_up, dispatch_uid='users.user_signed_up')
def on_user_signed_up(sender, request, user, **kwargs):
    """
    Signal handler to be called when a given user has signed-up.
    """
    request.is_signup = True
    track_event(
        CATEGORY_SIGNUP_FLOW,
        ACTION_SIGN_UP,
        request.session.get('sociallogin_provider')
    )
    if switch_is_active('welcome_email'):
        ....

@receiver(user_logged_in, dispatch_uid='users.user_logged_in')
def on_user_logged_in(sender, request, user, **kwargs):
    """
    Signal handler to be called when a given user has signed-in.
    """
    if not hasattr(request, 'is_signup'):
        track_event(
            CATEGORY_SIGNUP_FLOW,
            ACTION_SIGN_IN,
            request.session.get('sociallogin_provider')
        )
```

What do you think of that approach?

Reply:
They're both hacks. Your solution and my solution. But A) your solution appears to resolve the flaw you found in mind and B) that `is_almost_same_dates()` trick never felt good. 
I'll just copy your solution but attempt to add some code comments to explain the reasoning. (similar to what I did to justify the `is_almost_same_dates` hack)